tools/buildsys/src/public/variant.rs

```diff
@@ -0,0 +1,412 @@
+    /// that variant-sensitive builds will rebuild if the variant changes.
+    ///
+    /// # Example
+    ///
+    /// Given a variant `aws-k8s-1.21`, if this function has been called in `build.rs`, then
+    /// all of the following conditional complition checks would evaluate to `true`.
+    ///
+    /// `#[cfg(variant = "aws-k8s-1.21")]`
+    /// `#[cfg(variant_platform = "aws")]`
+    /// `#[cfg(variant_runtime = "k8s")]`
+    /// `#[cfg(variant_family = "aws-k8s")]`
+    /// `#[cfg(variant_version = "1.21")]`
+    /// `#[cfg(variant_type = "general_purpose")]`
+    pub fn emit_cfgs(&self) {
+        Self::rerun_if_changed();
+        println!("cargo:rustc-cfg=variant=\"{}\"", self);
+        println!("cargo:rustc-cfg=variant_platform=\"{}\"", self.platform());
+        println!("cargo:rustc-cfg=variant_runtime=\"{}\"", self.runtime());
+        println!("cargo:rustc-cfg=variant_family=\"{}\"", self.family());
+        println!("cargo:rustc-cfg=variant_version=\"{}\"", self.version());
+        println!("cargo:rustc-cfg=variant_type=\"{}\"", self.variant_type());
+    }
+
+    fn parse<S: Into<String>>(value: S) -> Result<Self> {
+        let variant = value.into();
+        let mut parts = variant.split('-');
+        let platform = parts
+            .next()
+            .with_context(|| error::VariantPartSnafu {
+                part_name: "platform",
+                variant: variant.clone(),
+            })?
+            .to_string();
+        ensure!(
+            !platform.is_empty(),
+            error::VariantPartEmptySnafu {
+                part_name: "platform",
+                variant: variant.clone()
+            }
+        );
+        let runtime = parts
+            .next()
+            .with_context(|| error::VariantPartSnafu {
+                part_name: "runtime",
+                variant: variant.clone(),
+            })?
+            .to_string();
+        ensure!(
+            !runtime.is_empty(),
+            error::VariantPartEmptySnafu {
+                part_name: "runtime",
+                variant: variant.clone()
+            }
+        );
+        let variant_family = format!("{}-{}", platform, runtime);
+        let variant_version = parts.next().map(|s| s.to_string());
+        if let Some(value) = variant_version.as_ref() {
+            ensure!(
+                !value.is_empty(),
+                error::VariantPartEmptySnafu {
+                    part_name: "variant_version",
+                    variant: variant.clone()
+                }
+            );
+        }
+        let variant_type = parts.next().map(|s| s.to_string());
+        if let Some(value) = variant_type.as_ref() {
+            ensure!(
+                !value.is_empty(),
+                error::VariantPartEmptySnafu {
+                    part_name: "variant_type",
+                    variant: variant.clone()
+                }
+            );
+        }
```

Comment:
Could you save yourself these two checks by adding a regex with the valid format of the variant? If `aws-k8s-` is given, the regex will catch the problem and the `is_empty()` check can be avoided.

Reply:
I would prefer not to use regex just to check if a tuple is empty.